graph/pull_v2.go

```diff
@@ -0,0 +1,361 @@
+		logrus.Debugf("Error getting v2 registry: %v", err)
+		return true, err
+	}
+
+	if err := p.pullV2Repository(repo, image, tag, imagePullConfig, sf); err != nil {
+		if registry.ContinueOnError(err) {
+			logrus.Debugf("Error trying v2 registry: %v", err)
+			return true, err
+		}
+		return false, err
+	}
+	return false, nil
+}
+
+func (p *v2Puller) pullV2Repository(repo distribution.Repository, image, tag string, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (err error) {
+	var tags []string
+	taggedName := image
+	if len(tag) > 1 {
+		tags = []string{tag}
+		taggedName = utils.ImageReference(image, tag)
+	} else {
+		var err error
+		tags, err = repo.Manifests().Tags()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	c, err := p.poolAdd("pull", taggedName)
+	if err != nil {
+		if c != nil {
+			// Another pull of the same repository is already taking place; just wait for it to finish
+			sf.FormatStatus("", "Repository %s already being pulled by another client. Waiting.", image)
+			<-c
+			return nil
+		}
+		return err
+	}
+	defer p.poolRemove("pull", taggedName)
+
+	var pulledNew bool
+	for _, tag := range tags {
+		pulledNew, err = p.pullV2Tag(repo, image, tag, taggedName, imagePullConfig.OutStream, sf)
+		if err != nil {
+			return err
+		}
+	}
+
+	WriteStatus(taggedName, imagePullConfig.OutStream, sf, pulledNew)
+
+	return nil
+}
+
+func (p *v2Puller) pullV2Tag(repo distribution.Repository, localName, tag, taggedName string, out io.Writer, sf *streamformatter.StreamFormatter) (bool, error) {
+	logrus.Debugf("Pulling tag from V2 registry: %q", tag)
+
+	// downloadInfo is used to pass information from download to extractor
+	type downloadInfo struct {
+		img     *image.Image
+		tmpFile *os.File
+		digest  digest.Digest
+		layer   distribution.ReadSeekCloser
+		size    int64
+		err     chan error
+	}
+
+	manifest, err := repo.Manifests().GetByTag(tag)
+	if err != nil {
+		return false, err
+	}
+	verified, err := p.validateManifest(manifest, tag)
+	if err != nil {
+		return false, fmt.Errorf("error verifying manifest keys: %v", err)
+	}
+	if verified {
+		// TODO(tiborvass): isn't this misleading since the layers have not been verified yet?
+		logrus.Printf("Image manifest for %s has been verified", taggedName)
+	}
+
+	out.Write(sf.FormatStatus(tag, "Pulling from %s", repo.Name()))
+
+	downloads := make([]downloadInfo, len(manifest.FSLayers))
+	for i := len(manifest.FSLayers) - 1; i >= 0; i-- {
+		img, err := image.NewImgJSON([]byte(manifest.History[i].V1Compatibility))
+		if err != nil {
+			logrus.Debugf("error getting image v1 json: %v", err)
+			return false, err
+		}
+		downloads[i].img = img
+		downloads[i].digest = manifest.FSLayers[i].BlobSum
+
+		// Check if exists
+		if p.graph.Exists(img.ID) {
+			logrus.Debugf("Image already exists: %s", img.ID)
+			continue
+		}
+
+		out.Write(sf.FormatProgress(stringid.TruncateID(img.ID), "Pulling fs layer", nil))
+
+		downloadFunc := func(di *downloadInfo) error {
```

Comment:
Breaking this out would suffer from the same problem breaking out every other function has had, there are too many variables which need to get passed along. This likely needs to be refactored into a manager which can hold all the extra state related to communicating progress and output.